pkg/agent/controller/networkpolicy/networkpolicy_controller_test.go

```diff
@@ -436,3 +456,102 @@ func TestAddNetworkPolicyWithMultipleRules(t *testing.T) {
 	assert.Equal(t, 2, controller.GetAddressGroupNum())
 	assert.Equal(t, 1, controller.GetAppliedToGroupNum())
 }
+
+func TestNetworkPolicyMetrics(t *testing.T) {
+	// Initialize NetworkPolicy metrics (prometheus)
+	metrics.InitializeNetworkPolicyMetrics()
+	controller, clientset, reconciler := newTestController()
+
+	addressGroupWatcher := watch.NewFake()
+	appliedToGroupWatcher := watch.NewFake()
+	networkPolicyWatcher := watch.NewFake()
+	clientset.AddWatchReactor("addressgroups", k8stesting.DefaultWatchReactor(addressGroupWatcher, nil))
+	clientset.AddWatchReactor("appliedtogroups", k8stesting.DefaultWatchReactor(appliedToGroupWatcher, nil))
+	clientset.AddWatchReactor("networkpolicies", k8stesting.DefaultWatchReactor(networkPolicyWatcher, nil))
+
+	protocolTCP := v1beta1.ProtocolTCP
+	port := intstr.FromInt(80)
+	services := []v1beta1.Service{{Protocol: &protocolTCP, Port: &port}}
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	go controller.Run(stopCh)
+
+	// Test adding policy1 with a single rule
+	policy1 := newNetworkPolicy("policy1", []string{"addressGroup1"}, []string{}, []string{"appliedToGroup1"}, services)
+	networkPolicyWatcher.Add(policy1)
+	networkPolicyWatcher.Action(watch.Bookmark, nil)
+	select {
+	case ruleID := <-reconciler.updated:
+		t.Fatalf("Expected no update, got %v", ruleID)
+	case <-time.After(time.Millisecond * 100):
+	}
+	checkNetworkPolicyMetrics(t, controller)
+
+	// Test adding policy2 with multiple rules
+	policy2 := newNetworkPolicyWithMultipleRules("policy2", []string{"addressGroup2"}, []string{"addressGroup2"}, []string{"appliedToGroup2"}, services)
+	networkPolicyWatcher.Add(policy2)
+	select {
+	case ruleID := <-reconciler.updated:
+		t.Fatalf("Expected no update, got %v", ruleID)
+	case <-time.After(time.Millisecond * 100):
+	}
+	checkNetworkPolicyMetrics(t, controller)
+
+	// Test deleting policy1
+	networkPolicyWatcher.Delete(newNetworkPolicy("policy1", []string{}, []string{}, []string{}, nil))
+	select {
+	case ruleID := <-reconciler.updated:
+		t.Fatalf("Expected no update, got %v", ruleID)
+	case <-time.After(time.Millisecond * 100):
```

Comment:
@yktsubo I see that you switched back to timeout rather than using waituntil. Is this because of all the other tests using this timeout mechanism or any other reason?

Reply:
Hi @srikartati 
Yes, the basic reason is that other test cases use the same logic.
This logic use waiting for a message from reconciler. So we don't need to poll it and this will stop after 100ms timeout, which is the same timeout of polling logic. So I think this logic is better than polling.
Actually I couldn't use reconciler in the integration test, so I used polling with timeout instead.